### PR TITLE
fix: ensure sequence metadata list accesses are threadsafe in task runner

### LINKED ISSUE #19458: java.lang.ArrayIndexOutOfBoundsException in streaming supervisor index task
### Affected Version

All Druid versions containing stream supervisor.

### Description

Druid stream supervisor task runner code has racy logic with the `sequences` copy-on-write list which is accessed by both the main thread (thread doing the processing) and the publishing thread (committing the segments). These two threads can race on the `sequences` list and cause the task to subsequently crash with the following exception:

```
java.lang.ArrayIndexOutOfBoundsException: Index 1 out of bounds for length 1
        at java.base/java.util.concurrent.CopyOnWriteArrayList.elementAt(CopyOnWriteArrayList.java:385) ~[?:?]
        at java.base/java.util.concurrent.CopyOnWriteArrayList.get(CopyOnWriteArrayList.java:398) ~[?:?]
        at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner.getLastSequenceMetadata(SeekableStreamIndexTaskRunner.java:1302) ~[druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner.runInternal(SeekableStreamIndexTaskRunner.java:631) ~[druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTaskRunner.run(SeekableStreamIndexTaskRunner.java:295) [druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.seekablestream.SeekableStreamIndexTask.runTask(SeekableStreamIndexTask.java:159) [druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.common.task.AbstractTask.run(AbstractTask.java:179) [druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:478) [druid-indexing-service-34.0.0.jar:34.0.0]
        at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:450) [druid-indexing-service-34.0.0.jar:34.0.0]
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131) [guava-32.1.3-jre.jar:?]
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76) [guava-32.1.3-jre.jar:?]
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82) [guava-32.1.3-jre.jar:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) [?:?]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) [?:?]
        at java.base/java.lang.Thread.run(Thread.java:840) [?:?]
```

### PR DESCRIPTION
**Pull request by jtuglu1**
### Description

Fixes https://github.com/apache/druid/issues/19458 and cleans up SeekableStreamIndexTaskRunnerTest.java

Old code had racy logic which was performing size checks on COW`sequences` in multiple places then attempting to index into the sequences list. This size check could be invalidated by a publishing thread calling .remove() on the between the size() check and the index operation.

Race:

1. [main runner thread]: calls getLastSequenceMetadata() [here](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java#L537-L540).
2. [main runner thread]: in the definition, `sequences.size()` returns 2 [here](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java#L1288)
3. [publisher thread]: publisher thread [removes](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java#L1046) element from the sequences list.
4. [main runner thread]: `sequences.get()` throws OOB error due to stale read race.

#### Fix

While technically this issue can be solved by taking snapshots before doing multiple reads on the sequence list, that still doesn't prevent read/write inter-leavings that might cause temporary lapses in in-memory/on-disk state (especially if we crash). So, I guarded the sequence list with a reentrant lock. While we re-acquire this lock per record, running this under load/taking some flamegraphs did not add any noticeable overhead. I believe this is mainly because the lock is not frequently contended in the common case and bottlenecks in the ingestion code lie elsewhere. Switching to a rw lock is another option here but performance was the same between the two options. Will add the performance benchmarks soon.

#### Release note

Fix fatal race in streaming ingest task during segment publish.

<hr>

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 3 of 3 changed files.

---

This is an automated review by Codex GPT-5.5

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the updated code for correctness, edge cases, concurrency, lifecycle, and integration risks; no issues found.

Reviewed 3 of 3 changed files.

---

This is an automated review by Codex GPT-5.5

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java [hunk 1]
```diff
@@ -243,7 +243,8 @@ public enum Status
   protected volatile boolean pauseRequested = false;
   private volatile long nextCheckpointTime;
 
-  private volatile CopyOnWriteArrayList<SequenceMetadata<PartitionIdType, SequenceOffsetType>> sequences;
+  private final Lock sequencesLock = new ReentrantLock();
+  private List<SequenceMetadata<PartitionIdType, SequenceOffsetType>> sequences = new ArrayList<>();
```
Comment (gianm) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:249]: Can we use `@GuardedBy("sequencesLock")` here to enforce proper locking?

**Review by gianm — APPROVED** [review]
